backend/main.py

```diff
@@ -97,11 +91,73 @@ def solve():
     else:
         abort(403)
 
-@app.route('/register', methods=['POST'])
+@app.route('/login', methods = ['POST'])
+def login():
+    email = request.form.get('email')
+    attempt = request.form.get('password')
+
+    db_hostname = os.environ['COS_DB_HOST']
+    db_username = os.environ['COS_DB_USERNAME']
+    db_password = os.environ['COS_DB_PASSWORD']
+    db_name = os.environ['COS_DB_DATABASE']
+    conn = psycopg2.connect(host=db_hostname, user=db_username,
+                            password=db_password, dbname=db_name)
+    
+    cur = conn.cursor()
+    cur.execute("SELECT password, api_key FROM users WHERE email='{}'".format(email))
+    rows = cur.fetchall()
+
+    if len(rows) > 0:
+        if pbkdf2_sha256.verify(attempt, rows[0][0]):
+            return json.dumps({'status': 0, 'token': rows[0][1]})
+        else:
+            return json.dumps({'status': 1})
+    else:
+        return json.dumps({'status': 1})
+
+@app.route('/register', methods = ['POST'])
 def register():
-    """ register, under construction. """
-    username = request.form.get('username')
+    name = request.form.get('name')
     password = request.form.get('password')
+    email = request.form.get('email')
+    institution = request.form.get('institution')
+
+    if not (name and password and email and institution):
+        # not enough fields
+        return json.dumps({'status': 2})
+    
+    db_hostname = os.environ['COS_DB_HOST']
+    db_username = os.environ['COS_DB_USERNAME']
+    db_password = os.environ['COS_DB_PASSWORD']
+    db_name = os.environ['COS_DB_DATABASE']
+    conn = psycopg2.connect(host=db_hostname, user=db_username,
+                            password=db_password, dbname=db_name)
+    
+    cur = conn.cursor()
+    cur.execute("SELECT 1 FROM users WHERE email='{}'".format(email))
+    rows = cur.fetchall()
+
+    if len(rows) > 0:
+        # email already registered
+        return json.dumps({'status': 1})
+    else:
+        users_cols = ['name', 'email', 'institution', 'password', 'api_key']
+        pass_hashed = pbkdf2_sha256.encrypt(password, rounds=200000, salt_size=16)
+        api_key = os.urandom(64).encode('hex')
+        
+        cur.execute('INSERT INTO users ({}) VALUES ({})'.format(
+            ", ".join(users_cols), ", ".join(["%s"]*len(users_cols))),
+                    (name,
+                     email,
+                     institution,
+                     pass_hashed,
+                     api_key,
+                     ))
+        conn.commit()
+        cur.close()
+        conn.close()
```

Comment:
Can you abstract out this database connection to a stand-alone method and then catch possible exceptions. You can take a look at what I did in `log_query`.  In general, this is how database connection should be implemented. And this allows us using the web interface offline, too.